engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala

```diff
@@ -441,34 +443,47 @@ class IrToTruffle(
                 else Left(l)
               }
               .map(fOpt =>
-                // Register builtin iff it has not been automatically registered at an early stage
-                // of builtins initialization or if it is a static wrapper.
-                fOpt
-                  .filter(m => !m.isAutoRegister() || staticWrapper)
-                  .map { m =>
-                    if (staticWrapper) {
-                      // Static wrappers differ in the number of arguments by 1.
-                      // Therefore we cannot simply get the registered function.
-                      // BuiltinRootNode.execute will infer the right order of arguments.
-                      val bodyBuilder =
-                        new expressionProcessor.BuildFunctionBody(
-                          fn.arguments,
-                          fn.body,
-                          effectContext,
-                          true
-                        )
-                      new RuntimeFunction(
-                        m.getFunction.getCallTarget,
-                        null,
-                        new FunctionSchema(
-                          new Array[RuntimeAnnotation](0),
-                          bodyBuilder.args(): _*
-                        )
+                fOpt.map { m =>
+                  if (m.isAutoRegister) {
+                    val irFunctionArgumentsCount = fn.arguments.length
+                    val builtinArgumentsCount =
+                      m.getFunction.getSchema.getArgumentsCount
+                    if (irFunctionArgumentsCount != builtinArgumentsCount) {
+                      val irFunctionArguments =
+                        fn.arguments.map(_.name.name).mkString(",")
+                      val builtinArguments =
+                        m.getFunction.getSchema.getArgumentInfos
+                          .map(_.getName)
+                          .mkString(",")
+                      throw new CompilerError(
+                        s"Wrong number of arguments provided in the definition of builtin function ${cons.getName}.${methodDef.methodName.name}. " +
+                        s"[$irFunctionArguments] vs [$builtinArguments]"
                       )
-                    } else {
-                      m.getFunction
                     }
+                    val bodyBuilder =
+                      new expressionProcessor.BuildFunctionBody(
+                        fn.arguments,
+                        fn.body,
+                        effectContext,
+                        true
+                      )
+                    val builtinRootNode =
+                      m.getFunction.getCallTarget.getRootNode
+                        .asInstanceOf[BuiltinRootNode]
+                    builtinRootNode.setModuleName(moduleScope.getModule.getName)
```

Comment:
Can't this information be passed to the factory that creates the node?

Reply:
There's not enough information on that stage. The `BuiltinMethod` annotation contains an unqualified method name. There is also no information on whether the method is static or not (to derive the self-type). Although the `BuiltinType` annotation contains a qualified type name. But I'm not sure if we can correctly match builtin methods and types because there's not always a 1-1 correspondence (i.e. like with numbers subtypes)

Reply:
CCing @hubertp as the _creator of these annotations_ and author of `Builtins::readBuiltinTypes` and `Buitins::readBuiltinMethodsMethods`.

I see that `BuiltinTypes.metadata` contains
```
Date_Time:org.enso.interpreter.node.expression.builtin.date.DateTime:Standard.Base.Data.Time.Date_Time.Date_Time:EnsoDateTime
```
which represents:
- Enso type name
- fully qualified Java class name
- fully qualified name as available in stdlib
- underlying type name of the builtin

then I also see that `BuiltinMethods.metadata` contains:
```
Date.month:org.enso.interpreter.node.expression.builtin.date.MonthDateMethodGen:false:true
```
which represents:
- the language-level name of this method
-  fully qualified Java class name
- **is static** information
- is autoregistered information

> not enough information on that stage

Given the combined content of the `BuiltinMethods.metadata` and  `BuiltinTypes.metadata` I am a believer in having all (most of) the needed information already available...

> no information on whether the method is static or not

...that information seems to be present in `BuiltinMethods.metadata`.

Should any information be missing in `BuiltinMethods.metadata` and  `BuiltinTypes.metadata` files, then I'd rather add such information there than trying to supply it later via setters in `IrToTruffle` stage. Immutability creates better systems that are easier to reason about. Having [half-initialized objects floating around the system](http://wiki.apidesign.org/wiki/Singleton#Singletons_are_Pathological_Liars.5B3.5D) is just asking for troubles.

Please accept my apology for not providing this review sooner than this PR got merged.

Reply:
I'm not sure if this will cover static wrappers. But yes, in general that information is ready in the metadata files that are being read in static initializers.

Reply:
So if I understand that correctly, instead of providing the moduleName and typeName here via setters on `BuiltinRootNode`, you are suggesting to tweak [Builtins$LoadedBuiltinMethod::toFunction](https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java#L650) method (and any other affected locations in the code) such that it will create a `BuiltinRootNode` with already initialized moduleName and typeName? That sounds like a good idea.

> Immutability creates better systems that are easier to reason about.

Indeed. Let me also add a note that setters are twice as much problematic in the Truffle environment than in pure Java. What if `BuiltinRootNode.setModuleName` is used during runtime? Should we manually invalidate any assumptions? Is it safe not to call `CompilerDirectives.transferToInterpreterAndInvalidate()` in these setters? Should the `moduleName` and `typeName` fields be marked as `@CompilationFinal`? (They probably should, theoretically, `BuiltinRootNode` can be shared among multiple contexts and can be a compilation constant; therefore, these fields can also be compilation constants). etc ...

TL;DR; I agree with Jaroslav, and I favor a follow-up PR that removes these setters (or tracking it in a separate issue). The sooner we do it, the better.